### PR TITLE
[Feature](function) Support array_compact function

### PR DESCRIPTION
**Pull request by Yukang-Lian**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 33.94 seconds
 load time: 452 seconds
 storage size: 17180344230 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221110041315_clickbench_pr_43100.html

File: be/src/vec/functions/array/function_array_compact.h [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+        return true;
+    }
+
+    bool _execute_string(const IColumn& src_column, const ColumnArray::Offsets64& src_offsets,
+                         IColumn& dest_column, ColumnArray::Offsets64& dest_offsets,
+                         const NullMapType* src_null_map, NullMapType* dest_null_map) {
+        const ColumnString* src_data_concrete = reinterpret_cast<const ColumnString*>(&src_column);
+        if (!src_data_concrete) {
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (eldenmoon) [reply be/src/vec/functions/array/function_array_compact.h:162]: we could use this more generic function to handle cases which contain more complex types like `ARRAY<ARRAY<STRING>>`, et...

File: docs/en/docs/sql-manual/sql-functions/array-functions/array_compact.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+array_compact(arr)
+```
+
+#### Arguments
+
+`arr` — The array to inspect.
+
+#### Returned value
+
+The array without duplicate.
```
Comment (eldenmoon) [inline docs/en/docs/sql-manual/sql-functions/array-functions/array_compact.md:43]: The array without **continuous** duplicate

File: be/src/vec/functions/array/function_array_compact.h [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+        block.replace_by_position(result, std::move(dest_column_ptr));
+        return Status::OK();
+    }
+
+private:
+    template <typename ColumnType>
+    bool _execute_number(const IColumn& src_column, const ColumnArray::Offsets64& src_offsets,
+                         IColumn& dest_column, ColumnArray::Offsets64& dest_offsets,
```
Comment (eldenmoon) [inline be/src/vec/functions/array/function_array_compact.h:100]: why not we just use `ColumnArray's nested column` as input arguments, and then we could ignore nullable situation, the code is more like 
```
++src_pos;
++res_pos;
for (; src_pos < src_offset; ++src_pos) {
    if (nested_src->compareAt(src_pos - 1, src_pos, *nested_src, 1)) {
        nested_dst_column->insert_from(src_values, src_pos);
        ++res_pos;
    }
}
```

if `nested_dst_column` and `nested_src` is nullable, then insert_from will handle whether the value is null or not automatically， so you won't need to take care of the null values

Comment (eldenmoon) [reply be/src/vec/functions/array/function_array_compact.h:100]: the code  could be more clean and clear

Comment (Yukang-Lian) [reply be/src/vec/functions/array/function_array_compact.h:100]: Yeah, you are right and I will refactor this part of the code soon.

**Comment by eldenmoon** [comment]
please rebase

File: be/src/vec/functions/array/function_array_compact.h [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                                get_name(), block.get_by_position(arguments[0]).type->get_name()));
+        }
+
+        block.replace_by_position(result, std::move(dest_column_ptr));
+        return Status::OK();
+    }
+
+private:
+    template <typename ColumnType>
+    bool _execute_number(const IColumn& src_column, const ColumnArray::Offsets64& src_offsets,
```
Comment (eldenmoon) [inline be/src/vec/functions/array/function_array_compact.h:83]: what's the difference of `_execute_number` and `_execute_generic `?

**Review by eldenmoon — DISMISSED** [review]
LGTM

**Comment by eldenmoon** [comment]
need rebase

**Review by eldenmoon — APPROVED** [review]
LGTM

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**